test/extensions/filters/network/ssh/service_userauth_test.cc

```diff
@@ -0,0 +1,746 @@
+  ASSERT_EQ(r.message(), "missing AllowResponse in auth state");
+}
+
+TEST_F(UpstreamUserAuthServiceTest, OnServiceAcceptedEncodingFailure) {
+  AuthState state;
+  state.allow_response = std::make_unique<pomerium::extensions::ssh::AllowResponse>();
+  std::string username_too_long(wire::MaxPacketSize, 'A');
+  state.allow_response->set_username(username_too_long);
+  EXPECT_CALL(*transport_, authState())
+    .WillRepeatedly(ReturnRef(state));
+  auto streamId = "stream-id"_bytes;
+  EXPECT_CALL(*transport_, sessionId())
+    .WillOnce(ReturnRef(streamId));
+
+  auto r = service_->onServiceAccepted();
+  ASSERT_EQ(r.code(), absl::StatusCode::kAborted);
+  ASSERT_EQ(r.message(), "error encoding user auth request: message size too large");
+}
+
+TEST_F(UpstreamUserAuthServiceTest, HandleMessageUserAuthBannerChannelNormal) {
+  AuthState state;
+  state.channel_mode = ChannelMode::Normal;
+  EXPECT_CALL(*transport_, authState())
+    .WillOnce(ReturnRef(state));
+
+  wire::UserAuthBannerMsg banner{};
+  wire::test::populateFields(banner);
+  wire::Message msg{banner};
+
+  EXPECT_CALL(*transport_, forward(Eq(msg), FrameTags{}));
+
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_OK(r);
+}
+
+TEST_F(UpstreamUserAuthServiceTest, HandleMessageUserAuthBannerChannelHijacked) {
+  AuthState state;
+  state.channel_mode = ChannelMode::Hijacked;
+  EXPECT_CALL(*transport_, authState())
+    .WillOnce(ReturnRef(state));
+
+  wire::UserAuthBannerMsg banner{};
+  wire::test::populateFields(banner);
+  wire::Message msg{banner};
+
+  // The banner message should not be forwarded downstream.
+
+  auto r = service_->handleMessage(std::move(msg));
+  ASSERT_OK(r);
+}
+
+TEST_F(UpstreamUserAuthServiceTest, HandleMessageExtInfo) {
+  TestSshMessageDispatcher d;
+  service_->registerMessageHandlers(d);
+
+  wire::ExtInfoMsg ext_info{};
+  wire::test::populateFields(ext_info);
+
+  EXPECT_CALL(*transport_, updatePeerExtInfo(Eq(ext_info)));
+
+  auto r = service_->handleMessage(wire::Message{ext_info});
+  ASSERT_OK(r);
+
+  // It is an error to receive multiple ExtInfo messages during the same auth exchange.
+  r = service_->handleMessage(wire::Message{ext_info});
+  ASSERT_EQ(r.code(), absl::StatusCode::kFailedPrecondition);
+  ASSERT_EQ(r.message(), "unexpected ExtInfoMsg received");
+}
+
+TEST_F(UpstreamUserAuthServiceTest, HandleMessageAuthSuccessNoPeerExtInfo) {
+  TestSshMessageDispatcher d;
+  service_->registerMessageHandlers(d);
+
+  // We need to first call onServiceAccepted() to initialize some internal state.
+  AuthState state;
+  state.allow_response = std::make_unique<pomerium::extensions::ssh::AllowResponse>();
+  state.allow_response->set_username("example-username");
+  EXPECT_CALL(*transport_, authState())
+    .WillRepeatedly(ReturnRef(state));
+  auto streamId = "stream-id"_bytes;
+  EXPECT_CALL(*transport_, sessionId())
+    .WillOnce(ReturnRef(streamId));
+  EXPECT_CALL(*transport_, sendMessageToConnection(_))
```

Comment:
```suggestion
  EXPECT_CALL(*transport_, sendMessageToConnection(MSG(wire::UserAuthSuccessMsg, _))
```
This is a good idea to make sure you aren't accidentally matching the wrong message